### PR TITLE
store: remove stores that have no region before balance

### LINKED ISSUE #52313: Remove stores that have been pruned during the detection
## Enhancement
Before sending mpp tasks, tidb will detect the liveness of tiflash nodes. However, tidb also detect nodes that have been pruned from the cluster. Once the detection of one tiflash node fails. The failed tiflash node will be added to a list to be periodically detected. If many tiflash nodes have been pruned before, the number of detection request will be very large and affects the network of production environment. So we need to detect nodes that are valid in the cluster.

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #52313

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/store/copr/batch_coprocessor.go [hunk 1]
```diff
@@ -919,17 +928,16 @@ func buildBatchCopTasksCore(bo *backoff.Backoffer, store *kvStore, rangesForEach
 			}
 		}
 
-		var batchTasks []*batchCopTask
-		var regionIDsInOtherZones []uint64
-		var regionInfosNeedReloadOnSendFail []RegionInfo
-		storeTaskMap := make(map[string]*batchCopTask)
+		rpcCtxs := make([]*tikv.RPCContext, 0)
+		validTiflashStores := make([][]uint64, 0)
+		allValidTiflashStores := make(map[uint64]struct{}, 0)
```
Comment (windtalker) [inline pkg/store/copr/batch_coprocessor.go:933]: The names of `validTiFlashStores` and `allValidTiFlashStores` make me confused. How about rename it to `usedTiFlashStores` and `usedTiFlashStoresMap`

File: pkg/store/copr/batch_coprocessor.go [hunk 2]
```diff
@@ -821,10 +824,20 @@ func filterAllStoresAccordingToTiFlashReplicaRead(allStores []uint64, aliveStore
 
 // getAliveStoresAndStoreIDs gets alive TiFlash stores and their IDs.
 // If tiflashReplicaReadPolicy is not all_replicas, it will also return the IDs of the alive TiFlash stores in TiDB zone.
-func getAliveStoresAndStoreIDs(ctx context.Context, cache *RegionCache, ttl time.Duration, store *kvStore, tiflashReplicaReadPolicy tiflash.ReplicaRead, tidbZone string) (aliveStores *aliveStoresBundle) {
+// func getAliveStoresAndStoreIDs(ctx context.Context, cache *RegionCache, allTiFlashStores []*locate.Store, ttl time.Duration, store *kvStore, tiflashReplicaReadPolicy tiflash.ReplicaRead, tidbZone string) (aliveStores *aliveStoresBundle) {
```
Comment (windtalker) [inline pkg/store/copr/batch_coprocessor.go:827]: Remove this line

File: pkg/store/copr/batch_coprocessor.go [hunk 3]
```diff
@@ -821,10 +824,20 @@ func filterAllStoresAccordingToTiFlashReplicaRead(allStores []uint64, aliveStore
...
+func getAliveStoresAndStoreIDs(ctx context.Context, cache *RegionCache, allValidTiflashStoresMap map[uint64]struct{}, ttl time.Duration, store *kvStore, tiflashReplicaReadPolicy tiflash.ReplicaRead, tidbZone string) (aliveStores *aliveStoresBundle) {
 	aliveStores = new(aliveStoresBundle)
 	allTiFlashStores := cache.RegionCache.GetTiFlashStores(tikv.LabelFilterNoTiFlashWriteNode)
-	aliveStores.storesInAllZones = filterAliveStores(ctx, allTiFlashStores, ttl, store)
+
+	allValidTiflashStores := make([]*tikv.Store, 0)
```
Comment (windtalker) [inline pkg/store/copr/batch_coprocessor.go:832]:
```suggestion
	allValidTiflashStores := make([]*tikv.Store, 0, len(allValidTiFlashStoresMap))
```

Comment (windtalker) [reply pkg/store/copr/batch_coprocessor.go:832]: And better to rename `Valid` to `Used`

File: pkg/store/copr/batch_coprocessor.go (hunk 1 above)
Comment (xzhangxian1008) [reply pkg/store/copr/batch_coprocessor.go:933]: > The names of `validTiFlashStores` and `allValidTiFlashStores` make me confused. How about rename it to `usedTiFlashStores` and `usedTiFlashStoresMap`

okk

File: pkg/store/copr/batch_coprocessor.go [hunk 4]
```diff
@@ -819,12 +822,24 @@ func filterAllStoresAccordingToTiFlashReplicaRead(allStores []uint64, aliveStore
 	return
 }
 
+func getAllUsedTiflashStores(allTiFlashStores []*tikv.Store, allUsedTiflashStoresMap map[uint64]struct{}) []*tikv.Store {
```
Comment (windtalker) [inline pkg/store/copr/batch_coprocessor.go]: What happens if a store exists in `allUsedTiFlashStoresMap` not not exist in `allTiFlashStores`?

**Review by windtalker — APPROVED** [review]
LGTM

File: pkg/store/copr/batch_coprocessor.go [hunk 5]
```diff
@@ -53,6 +53,8 @@ import (
 
 const fetchTopoMaxBackoff = 20000
 
+var regionNotFoundFlagForTest = false
```
Comment (zimulala) [inline pkg/store/copr/batch_coprocessor.go:56]: Didn't we use this variable in PR?

Comment (xzhangxian1008) [reply pkg/store/copr/batch_coprocessor.go:56]: > Didn't we use this variable in PR?

I will delete it.

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**